### PR TITLE
Editor 2

### PR DESCRIPTION
**Pull request by rbxnk**
Closes #64.

### CODE REVIEW

File: VisualPinball.Unity/VisualPinball.Unity/Import/VpxImporter.cs [hunk 1]
```diff
@@ -154,34 +155,42 @@ private void ImportRenderObjects(IRenderable item, RenderObjectGroup rog, GameOb
...
+				case Table table:			ic = table.SetupGameObject(obj, rog); break;
+				case Trigger trigger:		ic = obj.AddComponent<TriggerBehavior>().SetData(trigger.Data); break;
+			}
+			// for convenience move item behavior to the top of the list
+			if (ic != null) {
+				int numComp = obj.GetComponents<MonoBehaviour>().Length;
+				for (int i = 0; i <= numComp; i++) {
+					UnityEditorInternal.ComponentUtility.MoveComponentUp(ic);
+				}
 			}
```
Comment (freezy) [inline VisualPinball.Unity/VisualPinball.Unity/Import/VpxImporter.cs:194]: Does this work? Is that the order of the `MonoBehaviors` in the inspector?

image

File: VisualPinball.Unity/VisualPinball.Unity/VPT/Gate/GateBehavior.cs [hunk 2]
```diff
@@ -18,5 +19,36 @@ protected override Engine.VPT.Gate.Gate GetItem()
...
+		{
+			data.Center = pos.ToVertex2Dxy();
+			data.Height = pos.z;
+		}
+
+		public override ItemDataTransformType EditorRotationType => ItemDataTransformType.OneD;
+		public override Vector3 GetEditorRotation()
+		{
+			return new Vector3(data.Rotation, 0f, 0f);
+		}
```
Comment (freezy) [inline VisualPinball.Unity/VisualPinball.Unity/VPT/Gate/GateBehavior.cs:38]: Just randomly commenting that you can also do `public override Vector3 GetEditorRotation() => new Vector3(data.Rotation, 0f, 0f)` for oneliners. No biggie but it gets more readable IMO.

File: VisualPinball.Unity/VisualPinball.Unity/VPT/IEditableItemBehavior.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+		Vector3 GetEditorPosition();
+		void SetEditorPosition(Vector3 pos);
+
+		ItemDataTransformType EditorRotationType { get; }
+		Vector3 GetEditorRotation();
+		void SetEditorRotation(Vector3 pos);
+
+		ItemDataTransformType EditorScaleType { get; }
+		Vector3 GetEditorScale();
+		void SetEditorScale(Vector3 pos);
```
Comment (freezy) [inline VisualPinball.Unity/VisualPinball.Unity/VPT/IEditableItemBehavior.cs:20]: Maybe commenting those would explain other readers how cool they are? :)

File: VisualPinball.Unity/VisualPinball.Unity/VPT/Ramp/RampBehavior.cs [hunk 4]
```diff
@@ -18,5 +19,28 @@ protected override Engine.VPT.Ramp.Ramp GetItem()
...
+				return;
+			}
+
+			var diff = pos.ToVertex3D().Sub(data.DragPoints[0].Vertex);
+			diff.Z = 0f;
+			data.DragPoints[0].Vertex = pos.ToVertex3D();
+			for (int i = 1; i < data.DragPoints.Length; i++) {
+				var pt = data.DragPoints[i];
+				pt.Vertex = pt.Vertex.Add(diff);
+			}
```
Comment (freezy) [inline VisualPinball.Unity/VisualPinball.Unity/VPT/Ramp/RampBehavior.cs:43]: This is very cool. AFAIK in VP you have to select all the drag points and only then you can move the entire object.

**Comment by rbxnk** [comment]
Hey thanks for all the feedback, yeah this wasn't intended to be totally comprehensive, there's definitively still some missing fields and drag point support is a huge TODO. I've made a ticket to track the remaining work here: https://github.com/freezy/VisualPinball.Engine/issues/65
Please feel free to add things there in case I forgot anything.
I'll try to address the specific inline comments tomorrow.

File: VisualPinball.Unity/VisualPinball.Unity/Import/VpxImporter.cs (hunk 1 above)
Comment (rbxnk) [reply VisualPinball.Unity/VisualPinball.Unity/Import/VpxImporter.cs:194]: Hmm it seemed to work for me, its certainly not a supported API, but its the only way I could find to do it progammatically. Do you have a VPX file that repros this I could test?

File: VisualPinball.Unity/VisualPinball.Unity.Editor/Inspectors/TransformInspector.cs [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+
+		// work around for scale handle weirdness
+		private float _scaleFactor = 1.0f;
+
+		protected virtual void OnEnable()
+		{
+			_transform = target as Transform;
+
+			bool useDefault = true;
+			foreach (var t in targets) {
```
[the remark this answers is not in the document]
Comment (rbxnk) [reply VisualPinball.Unity/VisualPinball.Unity.Editor/Inspectors/TransformInspector.cs:31]: corrected with null coalescing, i have an aversion to linq in general for bloat and garbage, just one of those game dev habits

**Pull request merged**